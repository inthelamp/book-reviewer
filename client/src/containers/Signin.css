@media all and (min-width: 480px) {
    .signin {
      padding: 60px 0;
    }
  
    .signin form {
      margin: 0 auto;
      max-width: 320px;
    }

    .signin p {
      margin-top: 1.5rem;
    }
  }