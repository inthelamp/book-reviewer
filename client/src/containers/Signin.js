import { useState } from 'react'
import { useDispatch } from 'react-redux'
import { Form, Button } from 'react-bootstrap'
import { NavLink, Redirect } from 'react-router-dom'
import axios from 'axios'
import jwt_decode from 'jwt-decode'
import Message from '../components/Message'
import { SignedIn } from '../features/Actions'
import './Signin.css'

const Signin = ( props ) => {
  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')
  const [message, setMessage] = useState('')
  const [messageStyle, setMessageStyle] = useState('')
  const [isSignedIn, setIsSignedIn] = useState(false)

  const dispatch = useDispatch()

  const addHours = ( date, hours ) => {
    return date.getTime() + (hours*60*60*1000)
  }

  const setCookie = ( cookieName, value ) => {
    const today = new Date()
    today.setTime(addHours(today, process.env.REACT_APP_COOKIE_SUSTAINING_HOURS))
    const expires = 'expires='+ today.toUTCString()
    document.cookie = cookieName + '=' + value + '; SameSite=strict; Secure;' + expires + '; path=/'
  }

  const saveUser = (token) => {
    if (token) {      
      try {
          const decoded = jwt_decode(token)
          dispatch(SignedIn(decoded.name, decoded.role))
      } catch (error) {
          console.log(error)
      }
    } 
  }

  const onSubmit = (e) => {
    e.preventDefault()

    axios
    .post(process.env.REACT_APP_SERVER_BASE_URL + '/users/signin', {email, password})
    .then((response) => {
        setCookie('token', response.data.accessToken)
        saveUser(response.data.accessToken)
        setIsSignedIn(true)
    })
    .catch ((error) => {
        setMessageStyle('error_message')      
        if (error.response) {
          setMessage(error.response.data.message)                
        } else {
          console.log('Error', error.message)
        }
    })
  }

  const isValid = () => {
    return email.length > 0 && password.length > 0
  }  

  if (isSignedIn) {
    return <Redirect to='/' />
  }

  return (
    <div className='signin'>
      <Form onSubmit={onSubmit}>
        <Form.Group size='lg' controlId='email'>
          <Form.Label>Email</Form.Label>
          <Form.Control
            autoFocus
            type='email'
            value={email}
            onChange={(e) => setEmail(e.currentTarget.value)}
        />
        </Form.Group>
        <Form.Group size='lg' controlId='password'>
          <Form.Label>Password</Form.Label>
          <Form.Control
            type='password'
            value={password}
            onChange={(e) => setPassword(e.currentTarget.value)}
          />
        </Form.Group>
        <Button size='lg' type='submit' disabled={!isValid()} > 
          Sign In
        </Button>
        <Message message={message} messageStyle={messageStyle}/>
        <p>New Book Reviewer?&nbsp;<NavLink to='/signup'>Create an account</NavLink></p>
      </Form>      
    </div>
  )
}

export default Signin