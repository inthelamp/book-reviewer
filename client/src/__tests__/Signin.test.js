import React from 'react'
import { mount } from 'enzyme'
import { BrowserRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { screen, render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import Signin from '../containers/Signin'


describe('Component: Signin', () => {
  it('rendering correctly', () => {  

      const tree = renderer
          .create(<Router><Signin /></Router>)
          .toJSON()
      expect(tree).toMatchSnapshot()
  })

  it('rendering basic fields', () => {
      render(<Router><Signin /></Router>)
      expect(
        screen.getByRole('textbox', { name: 'Email' }))
        .toBeInTheDocument()
      expect(
        screen.getByText(/password/i))
        .toBeInTheDocument()
      expect(
        screen.getByRole('button', { name: 'Sign In' }))
        .toBeInTheDocument()
  })

  describe('Testing input onChange function', () => {
      let wrapper

      beforeEach(() => {
        wrapper = mount(<Router><Signin /></Router>)
      })

      it('Email input', () => {
        wrapper.find('Input').at(0).simulate('change', { target: { name: 'email', value: 'abc@company.com' } })
        expect(wrapper.find('Input').at(0).props().value).toEqual('abc@company.com')
      })

      it('Password input', () => {
        wrapper.find('Input').at(1).simulate('change', { target: { name: 'password', value: 'abcdef' } })
        expect(wrapper.find('Input').at(1).props().value).toEqual('abcdef')
      })
  })

  describe('Testing onSubmit function', () => {
      let wrapper

      beforeEach(() => {
        wrapper = mount(<Router><Signin /></Router>)
      })
      
      afterEach(() => {
        jest.resetAllMocks()
      })

      it('check submit', () => {
        expect(wrapper.find('form')).toHaveLength(1)
        const formEventMocked = { preventDefault: jest.fn() }
        wrapper.find('form').simulate('submit', formEventMocked)
        expect(formEventMocked.preventDefault).toBeCalledTimes(1)
      })
  })
})