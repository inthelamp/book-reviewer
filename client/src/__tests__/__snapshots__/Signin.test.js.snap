// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Signin rendering correctly 1`] = `
<div
  className="signin"
>
  <form
    className=""
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
      size="lg"
    >
      <label
        className="form-label"
        htmlFor="email"
      >
        Email
      </label>
      <input
        autoFocus={true}
        className="form-control"
        id="email"
        name="email"
        onChange={[Function]}
        type="email"
        value=""
      />
    </div>
    <div
      className="mb-3"
      size="lg"
    >
      <label
        className="form-label"
        htmlFor="password"
      >
        Password
      </label>
      <input
        className="form-control"
        id="password"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <button
      className="btn btn-primary btn-lg"
      disabled={true}
      type="submit"
    >
      Sign In
    </button>
    
    <p>
      New Book Reviewer?Â 
      <a
        aria-current={null}
        href="/signup"
        onClick={[Function]}
      >
        Create an account
      </a>
    </p>
  </form>
</div>
`;
